zinc/src/main/scala/sbt/internal/inc/MixedAnalyzingCompiler.scala

```diff
@@ -134,11 +134,11 @@ final class MixedAnalyzingCompiler(
     val incSrc = config.sources.filter(include)
     val (javaSrcs, scalaSrcs) = incSrc.partition(MixedAnalyzingCompiler.javaOnly)
     logInputs(log, javaSrcs.size, scalaSrcs.size, outputDirs)
-    val isPickleJava = config.currentSetup.order == Mixed && config.incOptions.pipelining && javaSrcs.nonEmpty
+    val pickleJava = Incremental.isPickleJava(config.currentSetup.options.scalacOptions)
```

Comment:
What happened to the other conditions here?

Reply:
The other conditions do not matter, I think. We need to run through scalac regardless of the compile order. Now that pipelining effectively means "skip javac", we could consider putting that in CompileOrder actually as `OnlyScala` or something.

Reply:
This should only be true when it's mixed compilation, so it was right before.

Reply:
Compilation order is about the order in which scalac and javac are called. For example, there's a pure Java project in sbt and it was using `CompileOrder.JavaThenScala`. It still needs to _generate_ pickle for the downstream no?

Reply:
Not in my eyes.  What's the point in running scalac and generating a pickle-only jar when you've got the whole bytecode jar/dir already?  You just feed that downstream.

The sources compiled by scalac are determined by `Mixed` so with this change `-Ypickle-java` + `JavaThenScala` will cause scalac to run on 0 sources, which is wrong and I'm not even sure how that behaves.

Reply:
> Not in my eyes. What's the point in running scalac and generating a pickle-only jar when you've got the whole bytecode jar/dir already? You just feed that downstream.

In sbt/sbt, `testingProj` depends on `testAgentProj`. In this case `testAgentProj` is the pure Java subproject with `CompileOrder.JavaThenScala`. The ordering should NOT matter because there are only Java sources in that subproject.

At a first glance it might seem pointless to call scalac on `testAgentProj`, but it's actually not true. What we want to do is regardless of the compile order etc, we want to let the downstream _`testingProj`_ start the compilation early, and since `testingProj` contains Scala sources, they can make use of the early output (pickle JAR). So even though `testAgentProj` itself needs bytecode JAR for _its_ upstream (in this cases none), it can still contribute to the overall pipelining.

Reply:
That doesn't make sense.  With `JavaThenScala` the earliest output is the bytecode from compiling the java code, which for `testAgentProj` is all the code.  Calling scalac is just wasting time - just feed the bytecode jar to testingProj, it's ready.

Reply:
I guess in case when you only have Java sources we could think of a way of communicating to downstream that information, but in a mixed source case, wouldn't `JavaThenScala` still require pickle generation?

Reply:
Yes, but you still don't drop the extra conditions here around `-Ypickle-java`.  You compiled the java sources to bytecode, you compile the scala sources if there are any (`scalaSrcs.nonEmpty`), and you coordinate downstream to compile with the java bytecode and the scala pickle once the latter are ready.

Also:

> We need to run through scalac regardless of the compile order.

If you pass both java+scala sources to scalac under the opposite `ScalaThenJava` order you break the invariant that the Scala can't depend on the Java.

Reply:
`ScalaThenJava` is a fiction that exists only in Scaladoc though - https://github.com/sbt/zinc/issues/235.
Even `JavaThenScala`, it mostly denotes the ordering during full compilation, and it's not like during incremental cycles we evict the stuff from Scala.

Here's a diagram I drew out:
![compile order vs pipelining](https://user-images.githubusercontent.com/184683/90546571-4210f280-e158-11ea-91da-fa82664d8ee8.jpg)

Reply:
In the above, if `core` uses `JavaThenScala`, then it cannot do pipelining itself, but it can still make early output so `app` subproject can use `early/core.jar`.

Reply:
Even if ScalaThenJava means the same ordering as Mixed, there's still a distinction in what sources it compiles.